### ISSUE TITLE
Support the creation of Applications using Graph permission: Application.ReadWrite.OwnedBy

### ISSUE THREAD

**Issue by PirateBread** [terraform-providers/terraform-provider-azuread#116]
I am getting a 403 permission denied error when trying to create an AzureAd Application and service principal.

According to the Microsoft Azure AD documentation there is a specific permission which allows an service principal to create and manage it's own applications/spns.

> The Application.ReadWrite.OwnedBy permission allows the same operations as Application.ReadWrite.All except that the former allows these operations only on applications and service principals that the calling app is an owner of. Ownership is indicated by the owners navigation property on the target application or service principal resource.

>Allows the calling app to create other applications and service principals, and fully manage those applications and service principals (read, update, update application secrets and delete)...

>NOTE: Using the Application.ReadWrite.OwnedBy permission to call GET /applications to list applications will fail with a 403. Instead use GET servicePrincipals/{id}/ownedObjects to list the applications owned by the calling application.

reference: https://docs.microsoft.com/en-us/graph/permissions-reference

It looks to me that Terraform provider does not yet support this alternative and requires the full "Application.ReadWrite.All" permissions which grants full access to all applications.

It would be much better if we can support creating apps/spns using this alternative implementation.

In this way we can envisage a scenario where a Terraform SPN can be master of it's own domain, creating AzureAD applications and resources and assigning rights appropriately without requiring full directory.write. access.

**Comment by PirateBread** [comment]
Anyone give us a 👍 ?
Or is this feature not that interesting?

**Comment by jmcshane** [comment]
I have a service principal with `Application.ReadWrite.OwnedBy` permission and `Directory.Read.All` permission. It is still failing with this same issue. Do I also need to grant it `Application.Read.All`?

**Comment by PirateBread** [comment]
> I have a service principal with `Application.ReadWrite.OwnedBy` permission and `Directory.Read.All` permission. It is still failing with this same issue. Do I also need to grant it `Application.Read.All`?

As it currently stands you must provide Application All permissions in order to create service principles and applications. 

I am hoping it is possible for this provider to be able to support the owned.by permission in the future so that it can create and manage in a secure way.

**Comment by manicminer** [comment]
Hi @PirateBread, thanks for raising this! I've tried to register both an application and service principal whilst authenticating the provider using a service principal having only `Application.ReadWrite.OwnedBy` on the Azure AD Graph API (not Microsoft Graph, as we are not using it to manage applications currently), but I was unable to reproduce any error. My application and service principal were registered and on inspection the SP I used to authenticate was assigned as an owner.

It's possible this was fixed in the time since you originally created the issue. Are you still able to reproduce?
